src/inference/src/ie_core.cpp

```diff
@@ -501,12 +513,132 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
+                if (layout == InferenceEngine::Layout::NC || layout == InferenceEngine::Layout::NCDHW ||
+                    layout == InferenceEngine::Layout::NCHW || layout == InferenceEngine::Layout::NHWC ||
+                    layout == InferenceEngine::Layout::NDHWC) {
+                    if (1 != item.second->getTensorDesc()
+                                 .getDims()[0])  // do not reshape/re-batch originally batched networks
+                        IE_THROW(NotImplemented)
+                            << "Auto-batching does not reshape/re-batch originally batched networks!";
+                    else
+                        atLeastOneOutputIsBatched = true;
+                }
+            }
+            if (!atLeastOneInputIsBatched || !atLeastOneOutputIsBatched)
+                IE_THROW(NotImplemented)
+                    << "Auto-batching supports only networks featuring inputs/outputs with the batched layouts !";
+
+            auto deviceNameWithoutBatch =
+                !deviceNameWithBatchSize.empty() ? DeviceIDParser::getBatchDevice(deviceNameWithBatchSize) : deviceName;
+            unsigned int requests = 0;
+            unsigned int optimalBatchSize = 0;
+            if (deviceNameWithBatchSize.empty()) {
+                // batch size is not set explicitly via device name e.g. BATCH:GPU(4)
+                // let's query the optimal batch size
+                std::map<std::string, ie::Parameter> options;
+                options["MODEL_PTR"] = &network;
+                optimalBatchSize = GetCPPPluginByName(DeviceIDParser(deviceNameWithoutBatch).getDeviceName())
+                                       .get_metric(METRIC_KEY(OPTIMAL_BATCH), options)
+                                       .as<unsigned int>();
+                auto res =
+                    GetConfig(deviceNameWithoutBatch, CONFIG_KEY(PERFORMANCE_HINT_NUM_REQUESTS)).as<std::string>();
+                requests = PerfHintsConfig::CheckPerformanceHintRequestValue(res);
+                const auto& reqs = config_with_batch.find(CONFIG_KEY(PERFORMANCE_HINT_NUM_REQUESTS));
+                if (reqs != config_with_batch.end())
+                    requests = (unsigned int)PerfHintsConfig::CheckPerformanceHintRequestValue(reqs->second);
+                if (requests) {
+                    optimalBatchSize = std::max(1u, std::min(requests, optimalBatchSize));
+                }
+            }
+            if (optimalBatchSize > 1 || !deviceNameWithBatchSize.empty()) {
+                auto function = network.getFunction();
+                bool bDetectionOutput = false;
+                for (auto&& node : function->get_ops()) {
+                    auto isDetectionOutputParent = [](decltype(node)& nd) {
+                        for (size_t n = 0; n < nd->get_input_size(); n++) {
+                            if (!std::strcmp("DetectionOutput", nd->get_input_node_ptr(n)->get_type_info().name))
```

Comment:
It won't work if we have several versions of DO operations.
Can we use RTTI instead of `strcmp` with `type_info`?

Reply:
as we discussed in the chat, the versions (opsets) of the DetectionOutput doesn't matter